File: plugins/grok/scripts/lib/workspace.mjs
import fs from "node:fs";
import path from "node:path";

/**
 * Walk up from cwd to find the nearest git root, else return cwd.
 * @param {string} cwd
 */
export function resolveWorkspaceRoot(cwd) {
  let current = path.resolve(cwd || process.cwd());
  while (true) {
    if (fs.existsSync(path.join(current, ".git"))) {
      return current;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      return path.resolve(cwd || process.cwd());
    }
    current = parent;
  }
}

File: crates/grok-companion/src/workspace.rs
use std::path::{Path, PathBuf};

/// Walk up from `cwd` to find the nearest `.git` directory; else return `cwd`.
pub fn resolve_workspace_root(cwd: &Path) -> PathBuf {
    let mut current = cwd
        .canonicalize()
        .unwrap_or_else(|_| cwd.to_path_buf());

    loop {
        if current.join(".git").exists() {
            return current;
        }
        if !current.pop() {
            return cwd
                .canonicalize()
                .unwrap_or_else(|_| cwd.to_path_buf());
        }
    }
}
